### DISCUSSION TITLE
Weighted Edges and alternative search algorithms

### DISCUSSION THREAD

**Discussion by stheid** [HeWeMel/nographs#19] [Q&A] [answered]
I have a graph problem where i need to find the shortest path in terms of edge or node weight. It seems to me that the current A* implementation cannot deal with that. is that correct?

I also need some kind of a greedy search algorithm with lookahead.

Overall, my question is: is that in the scope of the library and would it make sense to create a pullrequest, or should i rather just plainly implement it myself.

**Comment by HeWeMel** [comment]
Thanks for you questions.

**Reply by stheid** [reply]
Thanks for youre fast reply

**Comment by HeWeMel** [comment, marked as answer]
Could you please add more information about the requirement "edge or node" weight? To illustrate why I ask, here are two examples:

Situation 1)  We search for a shortest path  in terms of the sum of the weights of the edges along the path. This can be done with the [Dijkstra shortest paths algorithm, as implemented in NoGraphs](https://nographs.readthedocs.io/en/latest/traversals.html#:~:text=Class%20nographs.TraversalShortestPaths). If, additionally, we have an admissible heuristic function, the [AStar](https://nographs.readthedocs.io/en/latest/traversals.html#:~:text=Class%20nographs.TraversalAStar) algorithm can be used. The linked sections of the tutorial also contain links to examples.

Situation 2) We search for a shortest path in terms of the sum of the costs assigned to the vertices along this path. This can be done with the same algorithms, if we define the weight of an edge by the costs of the vertex it leads to, see the [example here](https://nographs.readthedocs.io/en/latest/concept_and_examples.html#examples:~:text=a%20move%20to%20a%20neighbor%20has%20%E2%80%9Ccosts%E2%80%9D%20that%20depend%20on%20its%20position).

**Reply by stheid** [reply]
ok, i think i understand:

if the successorgenerator returns sequendes of tuples, the first is interpret as the next vertexs, the second as a weight of that edge.

**Reply by HeWeMel** [reply]
Yes, where an algorithm expects edges with weights, this is correct. You can find a longer explanation and examples [here](https://nographs.readthedocs.io/en/latest/graphs_and_adaptation.html#graphs-with-edge-attributes).

**Comment by HeWeMel** [comment]
What kind of "greedy search algorithm with lookahead." do you have in mind? Could you give an example, e.g., in form of a problem to be solved?

**Reply by stheid** [reply]
wirh greedy + lookahead i mean the following.

let lookahead ne l=2. starting in node v 
- look at all path extensions of length l in v
- select the path with minimal total edge weight
- go one edge in the direction of the optimal path extension
- repeat

the result is a solution path with hopefully minimal weight (after all, its just partially informed)

**Reply by HeWeMel** [reply]
I think this could be done.

The inner loop could be realized with the Dijkstra shortest paths algorithm: a) Use a temporary graph with an edge relation that comes from the real one but prunes the graph at depth *l* (see section about search-aware graphs how to access the depth from within a next_edges function) to avoid any deeper searches. b) Then, when the vertices reported by the algorithm first hit depth *l*, this is your optimal path extension (since vertices are reported in ascending order of paths length).

The outer loop is easy: Do the above described process, and then go one step in the direction of the first step of the locally optimal path.

(Of cause your approach could lead to a globally suboptimal result, and it might not even find your goal vertex)
